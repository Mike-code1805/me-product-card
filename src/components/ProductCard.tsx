import { useProduct } from '../hooks/useProduct';
import styles from '../styles/styles.module.css';
import React, { createContext, CSSProperties } from 'react';
import {
  InitialValues,
  onChangeArgs,
  ProductCartHandlers,
} from '../interfaces/products.interfaces';
import {
  Product,
  ProductContextProps,
} from '../interfaces/products.interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children?: (args: ProductCartHandlers) => JSX.Element | null;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    product,
    onChange,
    value,
    initialValues,
  });
  return (
    <Provider
      value={{
        counter,
        increaseBy,
        maxCount,
        product,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children!({
          count: counter,
          isMaxCountReached: isMaxCountReached,
          maxCount,
          product,
          reset,
          increaseBy,
        })}
      </div>
    </Provider>
  );
};
