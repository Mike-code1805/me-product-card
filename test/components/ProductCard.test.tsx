import React from 'react';
import renderer from 'react-test-renderer';
import { product2 } from '../data/product';
import { ProductCard } from '../../src/components/index';

describe('ProductCard', () => {
  test('should render product card component', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <h1>Hi World</h1>}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should increase the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {({ count, increaseBy }) => (
          <>
            <h1>Hi World</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(+1)}>+</button>
          </>
        )}
      </ProductCard>
    );
    let tree = wrapper.toJSON();
    expect(wrapper.toJSON()).toMatchSnapshot();
    renderer.act(() => {
      (tree as any).children[2].props.onClick();
    });
    tree = wrapper.toJSON();
    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
