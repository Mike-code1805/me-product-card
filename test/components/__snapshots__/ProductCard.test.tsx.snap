// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should increase the counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hi World
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +
  </button>
</div>
`;

exports[`ProductCard should render product card component 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hi World
  </h1>
</div>
`;
